### PR TITLE
[LoopFlatten] Use loop versioning when overflow can't be disproven

### PR DESCRIPTION
**Pull request by john-brawn-arm**
Implement the TODO in loop flattening to version the loop when we can't prove that the trip count calculation won't overflow.

### CODE REVIEW

**Comment by sjoerdmeijer** [comment]
Looks like a very good addition. 
Just a quick question. The pass is off by default, but have you benchmarked this? Just wanted to check if there are no unexpected results (e.g. the extra runtime checks regressing things).

**Comment by john-brawn-arm** [comment]
> Looks like a very good addition. Just a quick question. The pass is off by default, but have you benchmarked this? Just wanted to check if there are no unexpected results (e.g. the extra runtime checks regressing things).

This improves several EEMBC benchmarks on cortex-m processors (or rather undoes the regression caused by https://github.com/llvm/llvm-project/commit/e13bed4c5f3544c076ce57e36d9a11eefa5a7815 canonicalizing in a way that removes the inbounds qualifier on gep instructions), and doesn't cause any noticeable regressions in those benchmarks.

Running the llvm-test-suite benchmarks Intel desktop this patch causes no significant changes to the results. In fact it has no effect when compiling for a 64-bit target at all, as then overflow is handled by widening the induction variable from 32 to 64 bit. Compiling with -m32 it does cause changes to benchmarks, but it's all in initialization code that contributes a negligible amount to the runtime.

**Review by sjoerdmeijer — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/LoopFlatten.cpp [hunk 1]
```diff
@@ -926,18 +936,53 @@ static bool FlattenLoopPair(FlattenInfo &FI, DominatorTree *DT, LoopInfo *LI,
...
+      // If the trip count type isn't legal then it won't be possible to check
+      // for overflow using only a single multiply instruction, so don't
+      // flatten.
+      LLVM_DEBUG(
+          dbgs() << "Can't check overflow efficiently, not flattening\n");
+      return false;
+    }
+    LLVM_DEBUG(dbgs() << "Multiply might overflow, versioning loop\n");
+
+    // Version the loop. The overflow check isn't a runtime pointer check, so we
```
Comment (ostannard) [inline llvm/lib/Transforms/Scalar/LoopFlatten.cpp:961]: There is an assertion in `LoopVersioning::versionLoop` to make sure that some runtime checks are emitted. There is also some code in there which generates checks using SCEV, I'm guessing that's what allows this to work? If so, could you expand this comment to explain why that is guaranteed?

**Review by ostannard — APPROVED** [review]
LGTM

**Pull request merged**